File: app/src/main/java/miles/diary/data/rx/OkHttpObservable.java
package miles.diary.data.rx;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import java.io.IOException;
import java.io.Reader;

import miles.diary.util.Logg;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import rx.Observable;
import rx.Subscriber;

/**
 * Created by mbpeele on 3/8/16.
 */
public class OkHttpObservable<T> {

    private final String url;
    private final Class<T> clazz;
    private final Gson gson;

    public OkHttpObservable(Builder<T> builder) {
        url = builder.url;
        clazz = builder.tClass;
        gson = builder.gson;
    }

    public Observable<T> execute(final OkHttpClient client) {
        return Observable.create(new Observable.OnSubscribe<T>() {
            @Override
            public void call(Subscriber<? super T> subscriber) {
                subscriber.onStart();
                try {
                    Response response = client.newCall(new Request.Builder()
                            .url(url)
                            .build()).execute();

                    Reader reader = response.body().charStream();
                    subscriber.onNext(gson.fromJson(reader, clazz));
                    subscriber.onCompleted();

                    reader.close();
                } catch (IOException e) {
                    Logg.log(e);
                    subscriber.onError(e);
                } catch (JsonSyntaxException e1) {
                    subscriber.onError(e1);
                    Logg.log(e1);
                }
            }});
    }

    public static <L> Builder<L> builder(Class<L> lClass) {
        return new Builder<>(lClass);
    }

    public static class Builder<T> {

        private String url;
        private Class<T> tClass;
        private Gson gson;

        public Builder(Class<T> tClass) {
            this.tClass = tClass;
        }

        public Builder<T> url(String url) {
            this.url = url;
            return this;
        }

        public Builder<T> gson(Gson gson) {
            this.gson = gson;
            return this;
        }

        public OkHttpObservable<T> build() {
            return new OkHttpObservable<>(this);
        }
    }
}

File: app/src/main/java/miles/diary/data/rx/OkHttpObservable.kt
package miles.diary.data.rx

import com.google.gson.Gson
import io.reactivex.Single
import io.reactivex.SingleSource
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.util.concurrent.Callable

/**
 * Created by mbpeele on 3/8/16.
 */
class OkHttpObservable<T> private constructor(builder: Builder<T>) {

    private val url: String?
    private val clazz: Class<T>?
    private val gson: Gson

    init {
        url = builder.url
        clazz = builder.tClass
        gson = builder.gson
    }

    fun execute(client: OkHttpClient): Single<T> {
        return Single.using(Callable<Response> {
            client.newCall(Request.Builder()
                    .url(url)
                    .build()).execute()
        }, Function<Response, SingleSource<T>> {
            val reader = it.body()!!.charStream()
            val responseObject = gson.fromJson<T>(reader, clazz!!)
            Single.just(responseObject)
        }, object : Consumer<Response> {
            override fun accept(response: Response) {
                response.body()?.close()
                response.close()
            }
        }, false)
    }

    class Builder<T> {

        internal var url: String? = null
        internal var tClass: Class<T>? = null
        internal var gson = Gson()

        fun target(tClass: Class<T>): Builder<T> {
            this.tClass = tClass
            return this
        }

        fun url(url: String): Builder<T> {
            this.url = url
            return this
        }

        fun gson(gson: Gson): Builder<T> {
            this.gson = gson
            return this
        }

        fun build(): OkHttpObservable<T> {
            return OkHttpObservable(this)
        }
    }

    companion object {

        fun <L> builder(): Builder<L> {
            return Builder()
        }
    }
}
